### PR TITLE
make sure that the scheduler and all heap allocations are aligned to a cache line boundary

### PR DESCRIPTION
**Pull request by stlankes**
- an alignment avoids false sharing
- using an exponential backoff algorithm in the spinning loop of the semaphore

### CODE REVIEW

**Comment by jschwe** [comment]
I checked out e683c86 and tested the SMP issue. The 4 core test reduced the runtime of the laplace test by 5 seconds (10%), however the runtime with 1-3 cores were more or less unchanged, so that could also be just noise (I only did one run).

**Comment by stlankes** [comment]
Currently, I try to solve issue #78. I assumed that we have false sharing in our system. However, the message `Trying to get the scheduler for core 1, but it isn't available` sounds crazy fro me. Ok, the access to SCHEDULER is a little bit dirty. I know that I insert only at boot time some elements. Consequently, I don't protect the data structure. Now, I use a Vector instead of a BTreeMap. But I except that a vector is a little bit faster, but doesn't solve the problem.

File: src/mm/allocator.rs [hunk 1]
```diff
@@ -11,14 +11,22 @@ use crate::mm::hole::{Hole, HoleList};
...
 const BOOTSTRAP_HEAP_SIZE: usize = 4096;
 
+#[cfg(target_arch = "x86_64")]
+const CACHE_LINE_SIZE: usize = 128;
+#[cfg(not(target_arch = "x86_64"))]
+const CACHE_LINE_SIZE: usize = 64;
+
 /// A fixed size heap backed by a linked list of free memory blocks.
+#[cfg_attr(target_arch = "x86_64", repr(align(128)))]
+#[cfg_attr(not(target_arch = "x86_64"), repr(align(64)))]
```
Comment (jbreitbart) [inline src/mm/allocator.rs:29]: Same as before, why do you align the struct to 128 on x86?

File: src/mm/allocator.rs [hunk 2]
```diff
@@ -97,12 +104,11 @@ impl Heap {
...
-			if size < HoleList::min_size() {
-				size = HoleList::min_size();
-			}
-			let size = align_up!(size, mem::align_of::<Hole>());
-			let layout = Layout::from_size_align(size, layout.align()).unwrap();
+			let mut size = cmp::max(layout.size(), HoleList::min_size());
+			size = align_up!(size, mem::align_of::<Hole>());
+			size = align_up!(size, CACHE_LINE_SIZE);
+			let layout =
+				Layout::from_size_align(size, cmp::max(layout.align(), CACHE_LINE_SIZE)).unwrap();
```
Comment (jbreitbart) [inline src/mm/allocator.rs:111]: Will this force a minimum alignmend of CACH_LINE_SIZE for every allocation? Do you know how this is done on Linux or any other OS?

File: src/mm/test.rs [hunk 3]
```diff
@@ -9,16 +9,23 @@
...
 	use std::prelude::v1::*;
 
 	use crate::mm::allocator::*;
 	use crate::mm::hole::*;
 
+	#[cfg(target_arch = "x86_64")]
+	const CACHE_LINE_SIZE: usize = 128;
+	#[cfg(not(target_arch = "x86_64"))]
+	const CACHE_LINE_SIZE: usize = 64;
+
```
Comment (jbreitbart) [inline src/mm/test.rs:24]: Is it possible to reuse the existing definitions?

File: src/mm/allocator.rs (hunk 1 above)
Comment (jbreitbart) [reply src/mm/allocator.rs:29]: Also can you reuse the CACHE_LINE_SIZE constant here?

File: src/scheduler/mod.rs [hunk 4]
```diff
@@ -49,6 +51,9 @@ impl SchedulerInput {
 		}
 	}
 }
+
+#[cfg_attr(target_arch = "x86_64", repr(align(128)))]
+#[cfg_attr(not(target_arch = "x86_64"), repr(align(64)))]
```
Comment (jbreitbart) [inline src/scheduler/mod.rs:56]: As before. I think the size is wrong and reusing the constant would make it easier to maintain and understand

**Review by jbreitbart — CHANGES_REQUESTED** [review]

File: src/mm/allocator.rs [hunk 5]
```diff
@@ -11,14 +11,22 @@ use crate::mm::hole::{Hole, HoleList};
...
+use core::cmp;
 use core::ops::Deref;
 use core::ptr::NonNull;
 use core::{mem, ptr};
 
 /// Size of the preallocated space for the Bootstrap Allocator.
 const BOOTSTRAP_HEAP_SIZE: usize = 4096;
 
+#[cfg(target_arch = "x86_64")]
+const CACHE_LINE_SIZE: usize = 128;
```
[the remark this answers is not in the document]
Comment (stlankes) [reply src/mm/allocator.rs:23]: Yes and no. The prefetched loads more than one cache line (https://github.com/crossbeam-rs/crossbeam/blob/master/crossbeam-utils/src/cache_padded.rs#L60-L70).

File: src/mm/test.rs (hunk 3 above)
Comment (stlankes) [reply src/mm/test.rs:24]: I don't known an existing definition.

File: src/scheduler/mod.rs (hunk 4 above)
Comment (stlankes) [reply src/scheduler/mod.rs:56]: See above

File: src/mm/allocator.rs (hunk 2 above)
Comment (stlankes) [reply src/mm/allocator.rs:111]: Hm, normally the try to find smallest power-of-two block size big enough to hold requested size plus header.

https://github.com/lattera/glibc/blob/895ef79e04a953cac1493863bcae29ad85657ee1/hurd/hurdmalloc.c#L213-L223

**Comment by stlankes** [comment]
Now, I am using crossbeam to align data.

File: src/mm/allocator.rs (hunk 1 above)
Comment (stlankes) [reply src/mm/allocator.rs]: I think that number in "align" must be a real number. You are not able to use a Rust constant.

File: src/mm/test.rs (hunk 3 above)
Comment (jbreitbart) [reply src/mm/test.rs]: I was referring to the ones you defined before

Comment (stlankes) [reply src/mm/test.rs]: Ah, yes, that is true. I will fix it

File: src/mm/allocator.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (jbreitbart) [reply src/mm/allocator.rs]: Well, no. The cache line size on x86 is 64 bytes. If you want to do padding on two cache lines that is fine, but then don't call it cache line size 😉

[the remark this answers is not in the document]
Comment (jbreitbart) [reply src/mm/allocator.rs]: Btw: polly calls it `hardware_destructive_interference_size`.

File: src/mm/allocator.rs [hunk 6]
```diff
@@ -11,14 +11,25 @@ use crate::mm::hole::{Hole, HoleList};
...
 use core::ptr::NonNull;
 use core::{mem, ptr};
 
 /// Size of the preallocated space for the Bootstrap Allocator.
 const BOOTSTRAP_HEAP_SIZE: usize = 4096;
 
+#[cfg(any(target_arch = "x86_64", target_arch = "aarch64"))]
+const CACHE_LINE_SIZE: usize = 128;
+#[cfg(not(any(target_arch = "x86_64", target_arch = "aarch64")))]
+const CACHE_LINE_SIZE: usize = 64;
```
Comment (jbreitbart) [inline src/mm/allocator.rs:25]: As said in the other discussion, at least the wording seems wrong

File: src/mm/test.rs [hunk 7]
```diff
@@ -9,16 +9,23 @@
...
 	use std::mem::{align_of, size_of};
 	use std::prelude::v1::*;
 
 	use crate::mm::allocator::*;
 	use crate::mm::hole::*;
 
+	#[cfg(any(target_arch = "x86_64", target_arch = "aarch64"))]
+	const CACHE_LINE_SIZE: usize = 128;
+	#[cfg(not(any(target_arch = "x86_64", target_arch = "aarch64")))]
+	const CACHE_LINE_SIZE: usize = 64;
```
Comment (jbreitbart) [inline src/mm/test.rs:23]: See before

**Review by jbreitbart — CHANGES_REQUESTED** [review]

File: src/mm/allocator.rs (hunk 1 above)
Comment (jbreitbart) [reply src/mm/allocator.rs]: You are right, didn't know that this is not possible.

File: src/mm/allocator.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (stlankes) [reply src/mm/allocator.rs]: Ok, I renamed it

File: src/mm/allocator.rs (hunk 2 above)
Comment (jbreitbart) [reply src/mm/allocator.rs]: Thanks for the link. That got me thinking. I am not sure why they do it they way they do it and I think jemalloc does it differently. Then I was thinking if we should try to do something that realloc can be done efficiently. After that I was worried that the implementation will get to complex and we'll have more errors than before. I guess, I am just not a fan of having to implement your own allocator, even though I understand why it is necessary at the moment.

Long story short: Feel free to ignore this comment and if you ever have good idea on how to get rid of implementing our own allocator I am all yours 😉

**Comment by stlankes** [comment]
@jbreitbart Do I answer all open points?

**Comment by stlankes** [comment]
bors try

**Comment by stlankes** [comment]
I assume that Jens agrees after his last changes.

**Review by jbreitbart — APPROVED** [review]
Fine with me :wink:
